arch/x86/entry/common.c:71:16: error: use of undeclared identifier '__noinline__' on upstream

When trying to extract symbols int80_emulation from arch/x86/entry/common.c:

```c
arch/x86/entry/common.c:71:16: error: use of undeclared identifier '__noinline__'
   71 | __attribute__((noinline)) __attribute__((no_instrument_function)) __attribute__((section(".noinstr.text"))) __attribute__((no_profile_instrument_function)) void int80_emulation(struct pt_regs *regs) {
      |                ^
/home/mpdesouza/git/linux/include/linux/compiler_attributes.h:245:56: note: expanded from macro 'noinline'
  245 | #define   noinline                      __attribute__((__noinline__))
      |                                                        ^
arch/x86/entry/common.c:151:48: error: declaration of anonymous struct must be a definition
  151 |     regs->orig_ax = regs->ax & ((((int)(sizeof(struct (unnamed struct at arch/x86/entry/common.c:246:29))))) + (((~(((0UL)))) - ((((1UL))) << (0)) + 1) & (~(((0UL))) >> (64 - 1 - (31)))));
      |                                                ^
arch/x86/entry/common.c:151:105: error: type name requires a specifier or qualifier
  151 |     regs->orig_ax = regs->ax & ((((int)(sizeof(struct (unnamed struct at arch/x86/entry/common.c:246:29))))) + (((~(((0UL)))) - ((((1UL))) << (0)) + 1) & (~(((0UL))) >> (64 - 1 - (31)))));
      |                                                                                                         ^
Error on pass: ClosurePass
In file included from arch/x86/entry/common.c:1:
In file included from arch/x86/entry/common.c:10:
In file included from /home/mpdesouza/git/linux/include/linux/kernel.h:16:
In file included from /home/mpdesouza/git/linux/include/linux/array_size.h:5:
/home/mpdesouza/git/linux/include/linux/compiler.h:5:1: warning: clang-extract: project #include's the same file multiple times. Only the first is registered.
    5 | #include <linux/compiler_types.h>
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

--- a/libcextract/Passes.cpp
+++ b/libcextract/Passes.cpp
@@ -86,7 +86,7 @@ static bool Build_ASTUnit(PassManager::Context *ctx, IntrusiveRefCntPtr<vfs::Fil
     return false;
   }
 
-  PrettyPrint::Set_Source_Manager(&AU->getSourceManager());
+  PrettyPrint::Set_AST(AU.get());
   ctx->AST = std::move(AU);
 
   return true;
@@ -171,16 +171,11 @@ class BuildASTPass : public Pass
     std::error_code ec;
     llvm::raw_fd_ostream out(Get_Dump_Name_From_Input(ctx), ec);
 
-    PrettyPrint::Set_Source_Manager(nullptr);
-    PrettyPrint::Set_Output_Ostream(&out);
-
     for (it = ctx->AST->top_level_begin(); it != ctx->AST->top_level_end(); ++it) {
       Decl *decl = *it;
       PrettyPrint::Print_Decl(decl);
     }
 
-    PrettyPrint::Set_Source_Manager(&ctx->AST->getSourceManager());
-    PrettyPrint::Set_Output_Ostream(nullptr);
     out.close();
   }
 
@@ -458,7 +453,7 @@ class FunctionExternalizerPass : public Pass
          and set its new SourceManager to the PrettyPrint class.  */
       ctx->AST->Reparse(std::make_shared<PCHContainerOperations>(),
                         ClangCompat_None, ctx->OFS);
-      PrettyPrint::Set_Source_Manager(&ctx->AST->getSourceManager());
+      PrettyPrint::Set_AST(ctx->AST.get());
 
       if (ctx->Ibt && ctx->AST) {
         /* Do a sanity check on IBT macros.  Some kernel branches can't use it,

--- a/libcextract/PrettyPrint.cpp
+++ b/libcextract/PrettyPrint.cpp
@@ -68,7 +68,7 @@ void PrettyPrint::Print_Decl_Raw(Decl *decl)
 
   /* If a SourceManager was specified and the decl source range seems valid,
      then output based on the original source code.  */
-  if (SM && Is_Range_Valid(decl_range)) {
+  if (Is_Range_Valid(decl_range)) {
 
     /* The range given by getSourceRange ignores attributes. Example :
 
@@ -215,28 +215,26 @@ void PrettyPrint::Print_RawComment(SourceManager &sm, RawComment *comment)
 
 StringRef PrettyPrint::Get_Source_Text(const SourceRange &range)
 {
-    /* Calling this function supposes that a SourceManager was given to this class.  */
-    assert(SM && "A SourceManager wasn't passed to PrettyPrint.");
-
     // NOTE: sm.getSpellingLoc() used in case the range corresponds to a macro/preprocessed source.
     // NOTE2: getSpellingLoc() breaks in the case where a macro was asigned to be expanded to typedef.
+    SourceManager &SM = AST->getSourceManager();
     auto start_loc = range.getBegin();//SM->getSpellingLoc(range.getBegin());
     auto last_token_loc = range.getEnd();//SM->getSpellingLoc(range.getEnd());
-    auto end_loc = clang::Lexer::getLocForEndOfToken(last_token_loc, 0, *SM, LangOpts);
+    auto end_loc = clang::Lexer::getLocForEndOfToken(last_token_loc, 0, SM, LangOpts);
     auto printable_range = clang::SourceRange{start_loc, end_loc};
     return Get_Source_Text_Raw(printable_range);
 }
 
 StringRef PrettyPrint::Get_Source_Text_Raw(const SourceRange &range)
 {
-    return clang::Lexer::getSourceText(CharSourceRange::getCharRange(range), *SM, LangOpts);
+    SourceManager &SM = AST->getSourceManager();
+    return clang::Lexer::getSourceText(CharSourceRange::getCharRange(range), SM, LangOpts);
 }
 
 /** Compare if SourceLocation a is before SourceLocation b in the source code.  */
 bool PrettyPrint::Is_Before(const SourceLocation &a, const SourceLocation &b)
 {
-  assert(SM && "No SourceManager were given");
-  BeforeThanCompare<SourceLocation> is_before(*SM);
+  BeforeThanCompare<SourceLocation> is_before(AST->getSourceManager());
 
   assert(a.isValid());
   assert(b.isValid());
@@ -246,15 +244,16 @@ bool PrettyPrint::Is_Before(const SourceLocation &a, const SourceLocation &b)
 
 void PrettyPrint::Debug_SourceLoc(const SourceLocation &loc)
 {
-  loc.dump(*SM);
+  loc.dump(AST->getSourceManager());
 }
 
 bool PrettyPrint::Contains_From_LineCol(const SourceRange &a, const SourceRange &b)
 {
-  PresumedLoc a_begin = SM->getPresumedLoc(a.getBegin());
-  PresumedLoc a_end   = SM->getPresumedLoc(a.getEnd());
-  PresumedLoc b_begin = SM->getPresumedLoc(b.getBegin());
-  PresumedLoc b_end   = SM->getPresumedLoc(b.getEnd());
+  SourceManager &SM = AST->getSourceManager();
+  PresumedLoc a_begin = SM.getPresumedLoc(a.getBegin());
+  PresumedLoc a_end   = SM.getPresumedLoc(a.getEnd());
+  PresumedLoc b_begin = SM.getPresumedLoc(b.getBegin());
+  PresumedLoc b_end   = SM.getPresumedLoc(b.getEnd());
 
   assert(a_begin.getFileID() == a_end.getFileID());
   assert(b_begin.getFileID() == b_end.getFileID());
@@ -292,8 +291,7 @@ bool PrettyPrint::Contains(const SourceRange &a, const SourceRange &b)
 /** Compare if SourceLocation a is after SourceLocation b in the source code.  */
 bool PrettyPrint::Is_After(const SourceLocation &a, const SourceLocation &b)
 {
-  assert(SM && "No SourceManager were given");
-  BeforeThanCompare<SourceLocation> is_before(*SM);
+  BeforeThanCompare<SourceLocation> is_before(AST->getSourceManager());
   return is_before(b, a);
 }
 
@@ -304,7 +302,7 @@ SourceLocation PrettyPrint::Get_Expanded_Loc(Decl *decl)
 
   /* If a SourceManager was specified and the decl source range seems valid,
      then output based on the original source code.  */
-  if (SM && Is_Range_Valid(decl_range)) {
+  if (Is_Range_Valid(decl_range)) {
 
     /* The range given by getSourceRange ignores attributes. Example :
 
@@ -319,11 +317,12 @@ SourceLocation PrettyPrint::Get_Expanded_Loc(Decl *decl)
 
     AttrVec &attrvec = decl->getAttrs();
     bool has_attr = false;
+    SourceManager &SM = AST->getSourceManager();
 
     for (size_t i = 0; i < attrvec.size(); i++) {
       const Attr *attr = attrvec[i];
       SourceLocation loc = attr->getRange().getEnd();
-      loc = SM->getExpansionLoc(loc);
+      loc = SM.getExpansionLoc(loc);
 
       if (loc.isValid() && Is_Before(furthest, loc)) {
         furthest = loc;
@@ -341,7 +340,7 @@ SourceLocation PrettyPrint::Get_Expanded_Loc(Decl *decl)
 
       /* Keep fetching tokens.  */
       while (true) {
-        auto maybe_next_tok = Lexer::findNextToken(head, *SM, LangOpts);
+        auto maybe_next_tok = Lexer::findNextToken(head, SM, LangOpts);
         Token *tok = ClangCompat_GetTokenPtr(maybe_next_tok);
 
         if (tok == nullptr) {
@@ -375,19 +374,20 @@ void PrettyPrint::Set_Output_To(const std::string &path)
 
 StringRef PrettyPrint::Get_Filename_From_Loc(const SourceLocation &loc)
 {
-  return SM->getFilename(loc);
+  return AST->getSourceManager().getFilename(loc);
 }
 
 OptionalFileEntryRef PrettyPrint::Get_FileEntry(const SourceLocation &loc)
 {
-  return SM->getFileEntryRefForID(SM->getFileID(loc));
+  SourceManager &SM = AST->getSourceManager();
+  return SM.getFileEntryRefForID(SM.getFileID(loc));
 }
 
 /* See PrettyPrint.hh for what they do.  */
 raw_ostream *PrettyPrint::Out = &llvm::outs();
 LangOptions PrettyPrint::LangOpts;
 PrintingPolicy PrettyPrint::PPolicy(LangOpts);
-SourceManager *PrettyPrint::SM;
+ASTUnit *PrettyPrint::AST;
 
 
 

--- a/libcextract/PrettyPrint.cpp
+++ b/libcextract/PrettyPrint.cpp
@@ -62,6 +62,18 @@ bool PrettyPrint::Is_Range_Valid(const SourceRange &range)
   return Is_Before(begin, end) && (begin != end);
 }
 
+bool Has_NoInline_Attr(Decl *decl)
+{
+  const AttrVec &attrs = decl->getAttrs();
+  for (const Attr *attr : attrs) {
+    if (attr->getKind() == attr::NoInline) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
 void PrettyPrint::Print_Decl_Raw(Decl *decl)
 {
   SourceRange decl_range = decl->getSourceRange();
@@ -122,7 +134,27 @@ void PrettyPrint::Print_Decl_Raw(Decl *decl)
           decl->print(Out, PPolicy);
         }
       } else {
+        MacroInfo *noinline_info = nullptr;
+        if (Has_NoInline_Attr(decl)) {
+          /* If the code has a `noinline` macro defined, it conflicts with
+             clang's __attribute__((noinline)) attribute dump.  Hence we
+             have to undef it.  */
+          Preprocessor &pp = AST->getPreprocessor();
+          MacroWalker MW(pp);
+
+          /* Get the valid macro at the source location.  */
+          IdentifierInfo *noinline = pp.getIdentifierInfo("noinline");
+          noinline_info = MW.Get_Macro_Info(noinline, decl->getBeginLoc());
+          /* Only print it if the noinline macro is defined at location.  */
+          if (noinline_info) {
+            Out << "#undef noinline\n";
+          }
+        }
         decl->print(Out, LangOpts);
+        if (noinline_info) {
+          /* Redeclare the macro to the previous value.  */
+          PrettyPrint::Print_MacroInfo(noinline_info);
+        }
       }
     } else {
       Out << decl_source;
@@ -192,7 +224,7 @@ void PrettyPrint::Debug_InclusionDirective(InclusionDirective *include)
 void PrettyPrint::Print_MacroInfo(MacroInfo *info)
 {
   SourceRange range(info->getDefinitionLoc(), info->getDefinitionEndLoc());
-  llvm::outs() << " info: " << Get_Source_Text(range) << '\n';
+  Out << "#define " << Get_Source_Text(range) << '\n';
 }
 
 void PrettyPrint::Print_Attr(Attr *attr)

--- a/libcextract/PrettyPrint.hh
+++ b/libcextract/PrettyPrint.hh
@@ -80,14 +80,14 @@ class PrettyPrint
 
   static bool Contains(const SourceRange &a, const SourceRange &b);
 
-  static inline void Set_Source_Manager(SourceManager *sm)
+  static inline void Set_AST(ASTUnit *ast)
   {
-    SM = sm;
+    AST = ast;
   }
 
   static inline SourceManager *Get_Source_Manager(void)
   {
-    return SM;
+    return &AST->getSourceManager();
   }
 
   static inline LangOptions &Get_Lang_Options(void)
@@ -153,9 +153,10 @@ class PrettyPrint
   /** Policy for printing.  We use the default for now.  */
   static PrintingPolicy PPolicy;
 
-  /** SourceManager built when parsing the AST.  Must be set after constructing
-      the ast by calling Set_Source_Manager.  */
-  static SourceManager *SM;
+  /** ASTUnit object.  Must be set after constructing the ast by
+      calling Set_Source_Manager.  FIXME: This should not be a statc
+      class variable.  Refactor this class entirely.  */
+  static ASTUnit *AST;
 
   friend class RecursivePrint;
 };

--- a/testsuite/small/attr-8.c
+++ b/testsuite/small/attr-8.c
@@ -1,0 +1,13 @@
+/* { dg-options "-DCE_EXTRACT_FUNCTIONS=func_f -DCE_NO_EXTERNALIZATION" }*/
+#define MACRO 0
+
+/* Do this sort of blurry definition to confuse out clang-extract into dumping
+   the AST of the function.  */
+#define DEFINE_FUNCTION(name) __attribute__((noinline)) int func_##name(void) { return MACRO; }
+#define DEFINE_VARIABLE(name) int var_##name;
+#define DEFINE(f) DEFINE_VARIABLE(f) DEFINE_FUNCTION(f)
+
+DEFINE(f)
+
+/* { dg-final { scan-tree-dump "__attribute__\(\(noinline\)\)" } } */
+/* { dg-final { scan-tree-dump-not "#define noinline" } } */

--- a/testsuite/small/attr-9.c
+++ b/testsuite/small/attr-9.c
@@ -1,0 +1,33 @@
+/* { dg-options "-DCE_EXTRACT_FUNCTIONS=func_f,g,h -DCE_NO_EXTERNALIZATION" }*/
+
+
+#define noinline SHOULD_NOT_BE_IN_THE_FINAL_OUTPUT
+
+#undef noinline
+#define noinline __attribute__((__noinline__))
+
+/* Do this sort of blurry definition to confuse out clang-extract into dumping
+   the AST of the function.  */
+#define DEFINE_FUNCTION(name) noinline int func_##name(void) { return 0; }
+#define DEFINE_VARIABLE(name) int var_##name;
+#define DEFINE(f) DEFINE_VARIABLE(f) DEFINE_FUNCTION(f)
+
+noinline int g()
+{
+  return 1;
+}
+
+DEFINE(f);
+
+noinline int h()
+{
+  return 2;
+}
+
+
+#undef noinline
+#define noinline SHOULD_NOT_BE_IN_THE_FINAL_OUTPUT
+
+/* { dg-final { scan-tree-dump "__attribute__\(\(noinline\)\) int func_f|int func_f\(void\) __attribute__\(\(noinline\)\)" } } */
+/* { dg-final { scan-tree-dump "#undef noinline" } } */
+/* { dg-final { scan-tree-dump-not "SHOULD_NOT_BE_IN_THE_FINAL_OUTPUT" } } */

